blog/2025-05-06-lifecycle-rules/index.mdx

```diff
@@ -0,0 +1,327 @@
+
+There’s a few problems with this however, the first one is that it’s essentially
+an infinitely growing folder and the second one is that statistically nobody
+will be looking at remote posts that are more than a week old. In ideal
+circumstances, the media cache is that, a cache. If the data doesn’t exist in
+the cache, your home server can always refetch the data and cache it again.
+
+This situation is where lifecycle rules can help a lot. With lifecycle rules you
+can set a policy that has objects start out in Standard, get punted down to
+Infrequent Access after a week, and then deleted after a month. This will save
+you a lot of money as your home server sees more and more posts.
+
+## How it works
+
+At its heart, this works off the back of our
+[global replication strategy](https://www.tigrisdata.com/blog/global-replication/).
+Every object belongs to a region, we replicate the objects around the world as
+needed. Otherwise, Tigris tries to fetch data in this order:
+
+![A diagram showing the storage access tiers: SSD cache is tried, then reading from FoundationDB, then reading from Regional Block Storage, then reading from Regional Block Stores in other regions](./cache-tiers.webp)
+
+We have a background job that checks over every object in every bucket with a
+lifecycle policy set. Every object in those buckets is automatically added to a
+secondary index by their creation time, so we do our lifecycle transitions in
+order. When the background job determines that it’s time for an object to be
+transitioned, it pushes the object’s internal ID to a queue. Another worker
+picks this up and then actually goes about moving the bytes around.
+
+From here it sounds like it’d be super easy, right? All you do is just copy and
+paste to the new location, update the metadata, and then you’re off to the
+races, right?
+
+Remember that SRE koan I shared earlier:
+
+<Xe top bottom>
+  The simpler something looks, the harder it is to make it happen.
+</Xe>
+
+This was hard for us because of our global replication system. Internally each
+object has metadata and the key bit of metadata that we use for replication is
+the Last-Modified timestamp. You can see this Last-Modified timestamp if you
+look at the response headers from the API:
+
+```text
+curl -s https://xe-demo-last-modified.fly.storage.tigris.dev/hello.txt --head | grep last-modified
+last-modified: Tue, 29 Apr 2025 15:47:02 GMT
+```
+
+Semantically, this Last-Modified header has _strong meaning_ to users. Users
+expect that if they modify the data in their object, it changes the
+Last-Modified header value. They also expect that if they DO NOT modify the
+data, the last-modified header value is not changed. Internally, this
+Last-Modified header has also been used to mean “the last time the metadata was
+modified” because the object metadata contains pointers to block storage.
+
+Most of the hard work we put into this was separating out the concepts of “last
```

Comment:
I suggest removing this. I am sure learning about manual testing won't make people comfortable with using Tigris. And I think this is a very very internal detail.